lib/policy_machine.rb

```diff
@@ -95,26 +95,39 @@ def is_privilege?(user_or_attribute, operation, object_or_attribute, options = {
   end
 
   ##
-  # Check the privilege without checking for prohibitions. May be called directly but is also used in is_privilege?
-  #
-  def is_privilege_ignoring_prohibitions?(user_or_attribute, operation, object_or_attribute, options = {})
-    unless user_or_attribute.is_a?(PM::User) || user_or_attribute.is_a?(PM::UserAttribute)
-      raise(ArgumentError, "user_attribute_pe must be a User or UserAttribute.")
-    end
+  # Can we derive a privilege given a set of filters?
+  def is_privilege_with_filters?(user_or_attribute, operation, object_or_attribute, filters: {}, options: {})
+    is_privilege_ignoring_prohibitions_with_filters?(user_or_attribute, operation, object_or_attribute, filters: filters, options: options) && (options[:ignore_prohibitions] || !is_privilege_ignoring_prohibitions?(user_or_attribute, PM::Prohibition.on(operation), object_or_attribute, options))
+  end
 
-    unless [PM::Operation, Symbol, String].any? { |allowed_type| operation.is_a?(allowed_type) }
-      raise(ArgumentError, "operation must be an Operation, Symbol, or String.")
-    end
+  ##
+  # Check the privilege with filters without checking for prohibitions.
+  def is_privilege_ignoring_prohibitions_with_filters?(user_or_attribute, operation, object_or_attribute, filters: {}, options: {})
+    assert_privilege_parameters!(user_or_attribute, operation, object_or_attribute)
 
-    unless object_or_attribute.is_a?(PM::Object) || object_or_attribute.is_a?(PM::ObjectAttribute)
-      raise(ArgumentError, "object_or_attribute must either be an Object or ObjectAttribute.")
+    if policy_machine_storage_adapter.respond_to?(:is_filtered_privilege?)
+      policy_machine_storage_adapter.is_filtered_privilege?(user_or_attribute, operation, object_or_attribute, filters: filters, options: options)
+    else
+      raise(NotImplementedError, "is_filtered_privilege? not implemented in storage adapter #{policy_machine_storage_adapter.class}")
     end
+  end
+
+  ##
+  # Check the privilege without checking for prohibitions. May be called directly but is also used in is_privilege?
+  #
+  def is_privilege_ignoring_prohibitions?(user_or_attribute, operation, object_or_attribute, options = {})
+    assert_privilege_parameters!(user_or_attribute, operation, object_or_attribute)
 
     if options.empty? && policy_machine_storage_adapter.respond_to?(:is_privilege?)
       privilege = [user_or_attribute, operation, object_or_attribute].map { |obj| obj.respond_to?(:stored_pe) ? obj.stored_pe : obj }
       return policy_machine_storage_adapter.is_privilege?(*privilege)
     end
 
+    if options[:filters] && policy_machine_storage_adapter.respond_to?(:is_privilege_with_filters?)
+      privilege = [user_or_attribute, operation, object_or_attribute].map { |obj| obj.respond_to?(:stored_pe) ? obj.stored_pe : obj }
```

Comment:
Why call this list of stored policy elements a privilege?

Reply:
This terminology mirrors the naming when calling `policy_machine_storage_adapter.is_privilege?` directly above. Additionally, there are several places throughout this file where a list of `[user/user_attribute, operation, object/object_attribute]` are referred to as a "privilege" - for example, where `PolicyMachine#privileges` and `PolicyMachine#scoped_privileges` are defined.